governance-chat/program/src/processor.rs

```diff
@@ -0,0 +1,114 @@
+//! Program processor
+
+use crate::{
+    error::GovernanceChatError,
+    instruction::GovernanceChatInstruction,
+    state::{assert_is_valid_chat_message, ChatMessage, GovernanceChatAccountType, MessageBody},
+    tools::account::create_and_serialize_account,
+};
+use borsh::BorshDeserialize;
+
+use solana_program::{
+    account_info::{next_account_info, AccountInfo},
+    clock::Clock,
+    entrypoint::ProgramResult,
+    msg,
+    program_error::ProgramError,
+    pubkey::Pubkey,
+    sysvar::Sysvar,
+};
+use spl_governance::state::{
+    governance::get_governance_data, proposal::get_proposal_data_for_governance,
+    token_owner_record::get_token_owner_record_data_for_realm,
+};
+
+/// Processes an instruction
+pub fn process_instruction(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    input: &[u8],
+) -> ProgramResult {
+    let instruction = GovernanceChatInstruction::try_from_slice(input)
+        .map_err(|_| ProgramError::InvalidInstructionData)?;
+
+    match instruction {
+        GovernanceChatInstruction::PostMessage { body } => {
+            msg!("GOVERNANCE-CHAT-INSTRUCTION: PostMessage");
+            process_post_message(program_id, accounts, body)
+        }
+    }
+}
+
+/// Processes PostMessage instruction
+pub fn process_post_message(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    body: MessageBody,
+) -> ProgramResult {
+    let account_info_iter = &mut accounts.iter();
+
+    let governance_program_info = next_account_info(account_info_iter)?; // 0
+    let governance_info = next_account_info(account_info_iter)?; // 1
+    let proposal_info = next_account_info(account_info_iter)?; // 2
+    let token_owner_record_info = next_account_info(account_info_iter)?; // 3
+    let governance_authority_info = next_account_info(account_info_iter)?; // 4
+
+    let chat_message_info = next_account_info(account_info_iter)?; // 5
```

Comment:
nit: this isn't a big deal because you're using `create_account` later, but you may want to explicitly check here that this account is actually `Uninitialized`